images/user/requirements.txt

```diff
@@ -0,0 +1,42 @@
+numpy
+scipy
+matplotlib
+pandas
+seaborn
+sympy
+networkx
+scikit-learn
+beautifulsoup4
+statsmodels
+nltk
+spacy
+geopandas
+# From https://github.com/utoronto-2i2c/jupyterhub-deploy/issues/29
+astroquery
+astropy
+galpy
+catalystcoop.pudl
+dask==2020.12.0
+distributed==2020.12.0
+dask-gateway==0.9.0
+# Infrastructure things
+jupyterlab==3.*
+voila==0.2.6
+nbgitpuller==0.9.*
+jupyter-resource-usage==0.5.0
+RISE==5.6.1
+jupyter_contrib_nbextensions==0.5.1
+notebook==6.1.4
+jupyterhub==1.3.0
+jupyter-rsession-proxy==1.2
+jupyter-tree-download==1.0.1
+ipywidgets==7.5.1
+otter-grader==2.1.3
+datascience
+geojson
+folium
+geopy
+# JupyterLab extensions
+jupyterlab_widgets==1.0.*
+jupytext==1.10.*
+jupyter-server-proxy==3.0.2
```

Comment:
Why we are not defining specific versions in this requirement file? I tend to always be explicit in the versions I am deploying on a base image, otherwise, you do not "control" which versions you are actually deploying... and that can introduce subtle bugs. I understand there is some additional work in keeping the version relevant as time pass but usually, that is a small price to pay, IMHO.

Reply:
I've seen it in other repos and used it myself too, but I don't know the exact answer to this question.

My intuition is that it's expected that an increase in the patch version to only fix bugs and don't cause any breaking change.
So pinning only the major and minor versions should be enough to prevent any issue caused by a pkg incompatibility.

Reply:
That's a sensible intuition. Although I should say I have been bitten by patch version incompatibilities before 😉 , this is why I am biased to be stricter on the version specification.